src/disposable/AsyncDisposable.ts

```diff
@@ -0,0 +1,42 @@
+import { IAsyncDisposable } from './types';
+
+export function createAsyncDisposable(func: () => void | Promise<void>): IAsyncDisposable {
+  return {
+    dispose: func,
+  };
+}
+
+export class AsyncDisposableCollection implements IAsyncDisposable {
+  protected readonly disposables: IAsyncDisposable[] = [];
+
+  get disposed(): boolean {
+    return this.disposables.length === 0;
+  }
+
+  public async dispose(): Promise<void> {
+    if (this.disposed) return;
+
+    while (!this.disposed) {
```

Comment:
```suggestion
    while (!this.disposed) {
```
the if statement feels redundant given the loop here.